File: 2022/05/05.py
from pythonfw.functions import extract_chunks

def preprocessing(input): 
    [a, b] = input.split('\n\n')
    rearrangements = extract_chunks(b, 3)
    cargo_lines = a.splitlines()
    cargo_width = len(cargo_lines[-1].replace(' ',''))
    cargo = ["" for _ in range(cargo_width)]
    for c in cargo_lines[:-1]: 
        v = c.split('[')
        s = len(v[0]) // 4
        for cc in v[1:]:
            cargo[s] =  cc[0] + cargo[s]
            s += 1 + (cc.count(' ') // 4)
    return (cargo, rearrangements)


def solver(procedure): 
    (cargo , rearrangements) = procedure
    cargos = {-1: list(cargo), 1: list(cargo)}
    for (l, s, e) in rearrangements:
        for version, cargo in cargos.items():
            cargo[e - 1] += cargo[s - 1][-l:][::version]
            cargo[s - 1] =  cargo[s - 1][:-l]

    for cargo in cargos.values():
        yield ''.join([stack[-1] for stack in cargo])

File: 2022/05/05.rs

aoc::main!(); 

fn preprocessing(input: &str) -> (Vec<String> , Vec<(usize, usize, usize)>) {
    let (crates, rearrangements) = input.split("\n\n").collect_tuple().unwrap() ;   
    let h = crates.lines().count() - 1  ;
    let w = (crates.len()-h) / (4 * h);
   (crates
        .replace("    ","_")
        .replace(['[',']',' '], "")
        .lines()
        .map(|l| format!("{}{}",l, "_".repeat(w - l.len())))
        .collect::<String>()
        .chars()
        .take(w * h)
        .enumerate()
        .sorted_by(|(m,_),(n,_)| (m % w).cmp(&(n % w)))
        .map(|(_, c)| c)
        .chunks(h)
        .into_iter()
        .map(|c| c.collect::<String>().replace("_", ""))
        .collect_vec() 
        ,

    rearrangements
        .lines()
        .map(|r| r.split_whitespace().filter_map(|s| s.parse().ok()).collect_tuple().unwrap())
        .map(|(n, start, end)| (n, start-1, end-1))
        .collect_vec())
}

fn part_1(input: (Vec<String> , Vec<(usize, usize, usize)>)) -> String {
    solver(input, |c| c.rev().collect::<String>())
}

fn part_2(input: (Vec<String> , Vec<(usize, usize, usize)>)) -> String {
    solver(input, |c| c.collect::<String>())
}

fn solver<F>(input: (Vec<String> , Vec<(usize, usize, usize)>) , f: F) -> String 
where
    F: Fn(std::string::Drain) -> String,
{
    let (mut stacks, _) = input ; 
    input.1
        .iter()
        .for_each(|&(n,s,e)| { 
            let crates: &str = &f(stacks[s].drain(..n));
            stacks[e].insert_str(0 , crates) ;
            });
    stacks
        .iter()
        .map(|c| c.chars().nth(0).unwrap())
        .collect::<String>()
}